crypto/signatures.go

```diff
@@ -23,25 +23,58 @@ type (
 var (
 	ErrNilInput         = errors.New("cannot use nil input")
 	ErrInvalidSignature = errors.New("invalid signature")
+	ErrRandUnexpected   = errors.New("unexpected result from random number generator")
 )
 
-// GenerateKeyPair creates a public-secret keypair that can be used to sign and
-// verify messages.
-func GenerateSignatureKeys() (sk SecretKey, pk PublicKey, err error) {
+// KeyPairGenerator is an interface that allows the caller to generate
+// public-secret key pairs.
+type KeyPairGenerator interface {
+	Generate() (SecretKey, PublicKey, error)
+	GenerateDetermistically(entropy [EntropySize]byte) (SecretKey, PublicKey)
+}
+
+// readBytesFunc is a function pointer that reads bytes into a buffer and
+// returns the total number of bytes written to the buffer.
+type readBytesFunc func([]byte) (int, error)
```

Comment:
might be good to use a convention like `randRead`

Also, we might want to export this, so that other packages can use it without needing to define their own.